package xspider.message;

message sskv_t {
    required string key = 1;
    //required string value = 2;
    optional string value = 2;
}

message msskv_t {
    required string key = 1;
    repeated sskv_t values = 2;
}

message sikv_t {
    required string key = 1;
    //required int32 value = 2;
    optional int32 value = 2;
}

message msikv_t {
    required string key = 1;
    repeated sikv_t values = 2;
}

message sfkv_t {
    required string key = 1;
    optional float value = 2;
}


message Message
{
    enum MSG_TYPE {
        PING = 1;
        PONG = 2;
        REQ_TASK = 3;
        REP_TASK = 4;
        TASK_SEED = 5;
        TASK_SEED_REP = 6;
        TASK_STATS = 7;
        TASK_STATS_REP = 8;
        ERROR = 9;
    }
    required MSG_TYPE type = 1;
    required uint64 ts = 2;
    optional bytes body = 3;
}

message Ping
{
    required string spider_id = 1;
    repeated sfkv_t status = 2;
}

message Pong
{
}


message TaskRequest 
{
    repeated sikv_t hosts_need = 1;
    repeated sikv_t hosts_exist = 2;
}

// msg.body zlib压缩
message TaskReply
{
    repeated string tasks = 1;
}

message TaskSeeds
{
    repeated string tasks = 1;
}

message TaskSeedsReply
{
    required string status = 1;
}

message TaskStats
{
}
message TaskStatsReply
{
    repeated sikv_t host_infos = 1;
}

